src/Messages/Attachments/TSAttachmentStream.m

```diff
@@ -289,6 +297,177 @@ + (void)deleteAttachments
     }
 }
 
+- (CGSize)calculateImageSize
+{
+    if ([self isVideo]) {
+        return [self videoThumbnail].size;
+    } else if ([self isImage] || [self isAnimated]) {
+        NSURL *_Nullable mediaUrl = [self mediaURL];
+        if (!mediaUrl) {
+            return CGSizeZero;
+        }
+
+        // With CGImageSource we avoid loading the whole image into memory.
+        CGImageSourceRef source = CGImageSourceCreateWithURL((CFURLRef)mediaUrl, NULL);
+        if (!source) {
+            OWSAssert(0);
+            return CGSizeZero;
+        }
+
+        NSDictionary *options = @{
+            (NSString *)kCGImageSourceShouldCache : @(NO),
+        };
+        NSDictionary *properties
+            = (__bridge_transfer NSDictionary *)CGImageSourceCopyPropertiesAtIndex(source, 0, (CFDictionaryRef)options);
+        CGSize imageSize = CGSizeZero;
+        if (properties) {
+            NSNumber *width = properties[(NSString *)kCGImagePropertyPixelWidth];
+            NSNumber *height = properties[(NSString *)kCGImagePropertyPixelHeight];
+            if (width && height) {
+                imageSize = CGSizeMake(width.floatValue, height.floatValue);
+            } else {
+                OWSAssert(0);
+            }
+        }
+        CFRelease(source);
+        return imageSize;
+    } else {
+        return CGSizeZero;
+    }
+}
+
+- (CGSize)ensureCachedImageSizeWithTransaction:(YapDatabaseReadWriteTransaction *_Nullable)transaction
+{
+    OWSAssert([NSThread isMainThread]);
+
+    if (self.cachedImageWidth && self.cachedImageHeight) {
+        return CGSizeMake(self.cachedImageWidth.floatValue, self.cachedImageHeight.floatValue);
+    }
+
+    CGSize imageSize = [self calculateImageSize];
+    self.cachedImageWidth = @(imageSize.width);
+    self.cachedImageHeight = @(imageSize.height);
+
+    void (^updateDataStore)() = ^(YapDatabaseReadWriteTransaction *transaction) {
+        OWSAssert(transaction);
+
+        NSString *collection = [[self class] collection];
+        TSAttachmentStream *latestInstance = [transaction objectForKey:self.uniqueId inCollection:collection];
+        if (latestInstance) {
+            latestInstance.cachedImageWidth = @(imageSize.width);
+            latestInstance.cachedImageHeight = @(imageSize.height);
+            [latestInstance saveWithTransaction:transaction];
+        } else {
+            // This message has not yet been saved; do nothing.
+            OWSAssert(0);
+        }
+    };
+
+    if (transaction) {
+        updateDataStore(transaction);
+    } else {
+        [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+            updateDataStore(transaction);
+        }];
+    }
+
+    return imageSize;
+}
+
+- (CGSize)cachedImageSizeWithTransaction:(YapDatabaseReadWriteTransaction *)transaction
+{
+    OWSAssert([NSThread isMainThread]);
+    OWSAssert(transaction);
+
+    if (self.cachedImageWidth && self.cachedImageHeight) {
```

Comment:
Isn't this redundant with the first few lines of `ensureCachedImageSizeWithTransaction`?

I also think the name of this method (cachedImageSizeWithTransaction) is confusing because you aren't necessarily getting a cached image size. I would expect a method named `cachedFoo` to return foo if it were cached, else return nil.

Can we just call `ensureCachedImageSizeWithTransaction` directly and delete this method? That would get rid of the redundancy and, I think be a better name.